test_util/test_upgrade_vpc.py

```diff
@@ -159,125 +132,277 @@ def wait_for_dns(hostname):
+        return [task['id'] for task in tasks]
+
+    def parse_dns_log(self, dns_log_content):
+        """Return a list of (timestamp, status) tuples from dns_log_content."""
+        dns_log = [line.strip().split(' ') for line in dns_log_content.strip().split('\n')]
+        if any(len(entry) != 2 or entry[1] not in ['SUCCESS', 'FAILURE'] for entry in dns_log):
+            message = 'Malformed DNS log.'
+            self.log.debug(message + ' DNS log content:\n' + dns_log_content)
+            raise Exception(message)
+        return dns_log
+
+    def log_test(self, test_name: str, call: Callable[[], None]) -> None:
+        try:
+            self.teamcity_msg.testStarted(test_name)
+            call()
+        except Exception:
+            # we want this except to be broad so that we can keep any exception from taking
+            # everything with it and not asserting the other tests
+            self.teamcity_msg.testFailed(test_name, details=traceback.format_exc())
+        finally:
+            self.teamcity_msg.testFinished(test_name)
+
+    @staticmethod
+    @retrying.retry(
+        wait_fixed=(1 * 1000),
+        stop_max_delay=(120 * 1000),
+        retry_on_result=lambda x: not x)
+    def wait_for_dns(dcos_api, hostname):
+        """Return True if Mesos-DNS has at least one entry for hostname."""
+        hosts = dcos_api.get('/mesos_dns/v1/hosts/' + hostname).json()
+        return any(h['host'] != '' and h['ip'] != '' for h in hosts)
+
+    def setup_cluster_workload(self, dcos_api: DcosApiSession, healthcheck_app: dict, dns_app: dict):
+        """Start a cluster workload on entry and verify its health on exit."""
+
+        # Deploy test apps.
+        # TODO(branden): We ought to be able to deploy these apps concurrently. See
+        # https://mesosphere.atlassian.net/browse/DCOS-13360.
+        with logger.scope("deploy apps"):
+            dcos_api.marathon.deploy_app(healthcheck_app)
+            dcos_api.marathon.ensure_deployments_complete()
+            # This is a hack to make sure we don't deploy dns_app before the name it's
+            # trying to resolve is available.
+            self.wait_for_dns(dcos_api, dns_app['env']['RESOLVE_NAME'])
+            dcos_api.marathon.deploy_app(dns_app, check_health=False)
+            dcos_api.marathon.ensure_deployments_complete()
+
+            test_apps = [healthcheck_app, dns_app]
+            self.test_app_ids = [app['id'] for app in test_apps]
+
+            self.tasks_start = {app_id: sorted(self.app_task_ids(dcos_api, app_id)) for app_id in self.test_app_ids}
+            self.log.debug('Test app tasks at start:\n' + pprint.pformat(self.tasks_start))
+
+            for app in test_apps:
+                assert app['instances'] == len(self.tasks_start[app['id']])
+
+    def verify_apps_state(self, dcos_api: DcosApiSession, dns_app: dict):
+        with logger.scope("verify apps state"):
+
+            # nested methods here so we can "close" over external state
```

Comment:
I see!  I think, this may not require as a separate comment. It is usually taken for granted that nested functions are utilized only for accessing objects from the immediate outer scope.